dotnet/sdk-phase-2.md

```diff
@@ -0,0 +1,484 @@
+  * Each task run will be synchronous on a single thread so deadlock detection can work
+  * All task scheduling should then work, e.g. `Task.Delay`, Dataflow stuff, etc
+* Follow Python and others' example of task ordering/distribution/etc
+* Not creating a `WorkflowRunner` abstraction like in Python because we have no reasonable expectation for a sandbox in
+  the future
+  * This abstraction can be created later, but since .NET does not have AppDomain anymore, there is no way to build a
+    good sandbox, so we will rely on custom analyzers instead
+
+Additions to `TemporalWorkerOptions`:
+
+```csharp
+namespace Temporalio.Worker;
+
+public class TemporalWorkerOptions : ICloneable
+{
+    // ... phase 1 pieces omitted ...
+
+    public IList<Type> Workflows { get; set; } = new List<Type>();
+
+    public TemporalWorkerOptions AddWorkflow(Type type) { /*... */ }
+
+    public int MaxCachedWorkflows { get; set; } = 10000;
+
+    public int MaxConcurrentWorkflowTasks { get; set; } = 100;
+
+    public double NonStickyToStickyPollRatio { get; set; } = 0.2;
+
+    public bool DisableRemoteActivities { get; set; }
+
+    public TimeSpan StickyQueueScheduleToStartTimeout { get; set; } = TimeSpan.FromSeconds(10);
+
+    internal static bool DebugModeEnvironmentVariable { get; } = /*...*/;
+
+    public bool DebugMode { get; set; } = DebugModeEnvironmentVariable;
+}
+```
+
+TODO(cretz): For discussion:
+
+* Intentionally leaving off `MaxConcurrentWorkflowTaskPolls`, right?
+
+## Workflow API
+
+```csharp
+namespace Temporalio.Workflows;
+
+public static class Workflow
+{
+    public static bool InWorkflow { get; }
+
+    public static WorkflowInfo Info { get; }
+
+    // Activities
+    // (assume all duplicated with "Local" before activity for local activities)
+
+    public static Task ExecuteActivityAsync(
+        Action activity, ActivityStartOptions options) { /*...*/ }
+
+    public static Task ExecuteActivityAsync<T>(
+        Action<T> activity, T arg, ActivityStartOptions options) { /*...*/ }
+
+    public static Task<TResult> ExecuteActivityAsync<TResult>(
+        Func<TResult> activity, ActivityStartOptions options) { /*...*/ }
+
+    public static Task<TResult> ExecuteActivityAsync<TResult>(
+        Func<Task<TResult>> activity, ActivityStartOptions options) { /*...*/ }
+
+    public static Task<TResult> ExecuteActivityAsync<T, TResult>(
+        Func<T, TResult> activity, T arg, ActivityStartOptions options) { /*...*/ }
+
+    public static Task<TResult> ExecuteActivityAsync<T, TResult>(
+        Func<T, Task<TResult>> activity, T arg, ActivityStartOptions options) { /*...*/ }
+
+    public static Task ExecuteActivityAsync(
+        string activity, object[] args, ActivityStartOptions options) { /*...*/ }
+
+    public static Task<TResult> ExecuteActivityAsync<TResult>(
+        string activity, object[] args, ActivityStartOptions options) { /*...*/ }
```

Comment:
I like how you avoided the ambiguity and only do multiple args with stringly typed activities.

Reply:
:+1: This was also done client side for workflows, signals, and queries. This was also done in Python (kinda, in a slightly different way).